### PR TITLE
Update the Project Structure and Raise Hotel Domain

### PR DESCRIPTION
**Pull request by DenysSemko**
## Why has it changed?

We need to add a new PR template.

## What has changed?

- Rearranged the project's structure
- Added a Hotel domain with the first CreatedEvent

## Ticket link
https://www.notion.so/Fix-Project-Structure-becba937e85e43be98e2581ede12ba5c

### CODE REVIEW

File: src/PrimeBooking.Domain/Informative/Hotel/Hotel.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        Name = name;
+        Capacity = capacity;
+        ContactInformation = contactInformation;
+        Facilities = facilities;
+        Stars = stars;
+    }
+
+    public Result Create()
+    {
+        if (string.IsNullOrEmpty(Name))
```
Comment (DenysSemko) [inline src/PrimeBooking.Domain/Informative/Hotel/Hotel.cs]: We need to add more validation cases here. 
As it's an aggregate root, we must treat them in one place. 
Proposals, @bochora96 please add more:
1. Name is null or empty
2. Capacity > 0
3. ContactInformation, regex for Email and Phone.
4. Address: not sure what we need to validate here.

File: src/PrimeBooking.Domain/Informative/Hotel/Errors/HotelErrors.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+using PrimeBooking.Domain.Informative.Common;
+
+namespace PrimeBooking.Domain.Informative.Hotel.Errors;
+
+public static class HotelErrors
```
Comment (DenysSemko) [inline src/PrimeBooking.Domain/Informative/Hotel/Errors/HotelErrors.cs:5]: Add ErrorFactory + extension for exceptions to build an error with it

File: src/PrimeBooking.Domain/Informative/Hotel/Hotel.cs (hunk 1 above)
Comment (DenysSemko) [reply src/PrimeBooking.Domain/Informative/Hotel/Hotel.cs]: Fixed

File: src/PrimeBooking.Domain/Informative/Hotel/Errors/HotelErrors.cs (hunk 2 above)
Comment (DenysSemko) [reply src/PrimeBooking.Domain/Informative/Hotel/Errors/HotelErrors.cs:5]: Fixed

File: src/PrimeBooking.Domain/Informative/Common/Error.cs [hunk 3]
```diff
@@ -0,0 +1,3 @@
+namespace PrimeBooking.Domain.Informative.Common;
+
+public record Error(string ErrorCode, string ErrorType, string Detail, string? StackTrace);
```
Comment (bochora96) [inline src/PrimeBooking.Domain/Informative/Common/Error.cs]: Do we need to add some enum to map it to proper status code? like enum StatusCode { NotFound, BadRequest, UnprocessableEntity etc. }

File: src/PrimeBooking.Domain/Informative/Common/Result.cs
Comment (bochora96) [inline src/PrimeBooking.Domain/Informative/Common/Result.cs:1]: Worth also to create some methods for Result like map, bind, match etc...

```
public static class ResultExtensions
{
    public static Result<T> Ensure<T>(
        this Result<T> result,
        Func<T, bool> predicate,
        Error error
    )
    {
        if (result.IsFailure)
        {
            return result;
        }

        return predicate(result.Value) ?
            result :
            Result.Failure<T>(error);
    }

    public static Result<TOut> Map<TIn, TOut>(
        this Result<TIn> result,
        Func<TIn, TOut> mappingFunc,
        ErrorPayload errorPayload
    )
    {
        return result.IsSuccess ?
            Result.Success(mappingFunc(result.Value)) :
            Result.Failure<TOut>(result.Errors, errorPayload);
    }

    public static async Task<Result> Bind<TIn>(
        this Result<TIn> result,
        Func<TIn, Task<Result>> func,
        Guid operationId
    )
    {
        if (result.IsFailure)
        {
            return Result.Failure(result.Errors);
        }

        return await func(result.Value);
    }

    public static async Task<Result<TOut>> Bind<TIn, TOut>(
        this Result<TIn> result,
        Func<TIn, Task<Result<TOut>>> func,
        ErrorPayload errorPayload = default
    )
    {
        if (result.IsFailure)
        {
            return Result.Failure<TOut>(result.Errors, errorPayload);
        }

        return await func(result.Value);
    }
}
```

File: src/PrimeBooking.Domain/Informative/Common/Error.cs (hunk 3 above)
Comment (DenysSemko) [reply src/PrimeBooking.Domain/Informative/Common/Error.cs]: Good idea, I will add the default one (HttpStatusCode)

File: src/PrimeBooking.Domain/Informative/Common/Result.cs
Comment (DenysSemko) [reply src/PrimeBooking.Domain/Informative/Common/Result.cs:1]: Agree!

File: src/PrimeBooking.Domain/PrimeBooking.Domain.csproj [hunk 4]
```diff
@@ -6,4 +6,12 @@
         <Nullable>enable</Nullable>
     </PropertyGroup>
 
+    <ItemGroup>
+      <Folder Include="Structural\" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <PackageReference Include="MediatR" Version="12.4.0" />
```
Comment (bochora96) [inline src/PrimeBooking.Domain/PrimeBooking.Domain.csproj]: As its pet project we can think of changing MediatR with Source Generated MediatR. Only for interest: https://www.youtube.com/watch?v=aaFLtcf8cO4

Comment (DenysSemko) [reply src/PrimeBooking.Domain/PrimeBooking.Domain.csproj]: Nice catch, uploaded new package

**Pull request merged**